contracts/utils/SelfMulticall.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./interfaces/ISelfMulticall.sol";
+
+/// @notice Contract that enables calls to the inheriting contract to be
+/// batched
+/// @dev Refer to OpenZeppelin's Multicall.sol for a similar implementation
+contract SelfMulticall is ISelfMulticall {
+    /// @notice Batches calls to the inheriting contract and reverts if at
+    /// least one of the batched calls reverts
+    /// @dev Bubbles up the raw returndata if a call reverts, leaving decoding
+    /// to the user
+    /// @param data Array of calldata of batched calls
+    /// @param returndata Array of returndata of batched calls
+    function multicall(bytes[] calldata data)
+        external
+        override
+        returns (bytes[] memory returndata)
+    {
+        returndata = new bytes[](data.length);
+        for (uint256 i = 0; i < data.length; i++) {
+            bool success;
+            (success, returndata[i]) = address(this).delegatecall(data[i]); // solhint-disable-line avoid-low-level-calls
+            // Adapted from OpenZeppelin's Address.sol
+            if (!success) {
+                if (returndata.length > 0) {
```

Comment:
Shouldn't this be `returndata[i]`?

Reply:
If yes, then maybe we need a new local memory variable and assign it to returndata[i] right at the end of this function. Something like: 
```
(bool success, bytes memory result) = address(this).delegatecall(data[i]);
if (!success) {
  ...
}
returndata[i] = result;
```

Reply:
> Shouldn't this be returndata[i]?

Nice catch

> If yes, then maybe we need a new local memory variable and assign it to returndata[i] right at the end of this function. 

I avoided that because it optimizes the non-happy path at the cost of making the happy path more expensive. In a non-try multicall we don't expect a revert so the non-happy path should be rarer.